---
title: "run_HUManN3"
author: "Jimmy Zhong"
date: "6/21/2022"
output: html_document
---

## Introduction

This is a demo to show the workflow to find multivariate
association in metagenomic sequencing data.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

We use *MetaPhlan2* to find the taxonomic composition of a metagenomic sample.
And *humann3*  to find the abundance of pathways of a metagenomic sample.

# https://huttenhower.sph.harvard.edu/metaphlan2/
Software MetaPhlan: "Metagenomic Phylogenetic Analysis"

# https://huttenhower.sph.harvard.edu/humann/
Software Humann: "HMP Unified Metabolic Analysis Network"

# https://github.com/biobakery/biobakery
Both tools are part of the *biobakery* workflow from the Huttenhower lab.
Follow the instruction on their Github to install biobakery3

```{bash}
https://github.com/biobakery/biobakery_workflows
conda create --name biobakery3 python=3.7
conda activate biobakery3
conda install -c biobakery biobakery_workflows

# Install HUMAnN 3.0 software with demo databases
# will also automatically install MetaPhlAn 3.0
conda install humann -c biobakery

# To upgrade your pangenome database:
humann_databases --download chocophlan full /localdata/researchdrive/zhongj2/biobakery_db --update-config yes
# To upgrade your protein database:
humann_databases --download uniref uniref90_diamond /localdata/researchdrive/zhongj2/biobakery_db --update-config yes
# To upgrade your annotations database:
humann_databases --download utility_mapping full /localdata/researchdrive/zhongj2/biobakery_db --update-config yes
```


## Run humann3 (which will also run MetaPhlan2)

Run it through each of your samples.
```{bash}
sbatch -p medium -c 10 --mem=40G -t 1-00:00 ~/scripts/run_humann3.bash ${out_folder_name} ${r1} ${r2} ${out_dir} 10
```

## Normalize the pathway and genefamilies abundance

This step is required to compare across samples. 
*relab" stands for "relative abundance", which is RPK (reads per kilobase)
```{bash}
humann_renorm_table --input SPARK019_pathabundance.tsv --units relab --output SPARK019_pathabundance.rel.tsv
humann_renorm_table --input SPARK019_genefamilies.tsv --units relab --output SPARK019_genefamilies.rel.tsv
```

## Merge the taxon and pathway abundance of each sample
```{bash}
merge_metaphlan_tables.py metaphlan/*_metaphlan_bugs_list.tsv > combined_metaphlan.tsv

humann_join_tables --input combined/genefamilies --output combined_genefamilies.tsv
humann_join_tables --input combined/pathabundance --output combined_pathabundance.tsv
humann_join_tables --input combined/pathcoverage --output combined_pathcoverage.tsv
```