---
title: "Tara Ocean metagenome COG and nutrient limitations"
author: "Jimmy Zhong"
date: "4/12/2023"
output: html_document
---

## Jimmy's ibaray set up
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# setting the working directory to the location of this file
setwd(dirname(rstudioapi::getActiveDocumentContext()$path ))
source(".init_tara.R")

tara_md = init_tara_metadata()
COGabun = get_COGabundance() # only the prokaryote fraction

tara_md = tara_md[colnames(COGabun), ]
tara_md$size_fraction = as.factor(tara_md$size_fraction)
```

# run MaAsLin2
```{r}
# result_dir = 'MaAsLin_out/logPO4_Salinity_Oxygen_Nitrogen_temp_logDepth_size'
# param = c('size_fraction','Mean_Temperature', "log_depth", 'Mean_Salinity',
#          'Mean_Oxygen', 'log_PO4', 'log_Nitrates', 'log_NO2', 'log_NO2NO3')

result_dir = 'MaAsLin_out/logPO4_logNO2NO3_Salinity_Oxygen_temp_logDepth_size'
param = c('size_fraction','Mean_Temperature', "log_depth", 'Mean_Salinity',
          'Mean_Oxygen', 'log_PO4','log_NO2NO3')

fit_data <- Maaslin2(
    COGabun, tara_md, result_dir, transform = "LOG",
    fixed_effects = param,
    random_effects = c("Mean_Lat", "Mean_Long"),
    reference = 'size_fraction,0.22-1.6',
    max_significance = 0.05,
    normalization = 'NONE',
    min_prevalence = 0.33, # at least 1/3 of the sample has it
    standardize = FALSE)

# tmp = COGabun[c("COG0226", "COG0573", "COG0581", "COG4985", "COG1117", "COG3638", "COG3221", "COG3639", "COG3624", "COG3625", "COG3626", "COG3627", "COG4107", "COG4778", "COG3454", "COG3709", "COG2764", "COG0395", "COG1653", "COG1134", "COG1392", "COG1840", "COG0704"), ]
# tmp$COG_function = c("PstS", "PstC", "PstA", "PstA_auxiliary", "PstB", "PhnC", "PhnD", "PhnE", "PhnG", "PhnH", "PhnI", "PhnJ", "PhnK", "PhnL", "PhnM", "PhnN", "PhnB", "UgpE", "UgpB", "TagH", "YkaA", "AfuA", "PhoU")
# tmp = tmp[c(ncol(tmp), 2:ncol(tmp)-1)]
# write.csv(tmp, "phosphorous_COGs_abundance.csv", row.names=TRUE)
```

# Screen for simple spearman correlation between phosphate/nitrate concentration and gene abundance
```{r}
return_corr2 = function(x, substrate) {
  y = tara_md %>% pull(substrate)
  obj = cor.test(x, y, method = "spearman")
  return(c(p.val = obj$p.value, corr = as.numeric(obj$estimate)))
}

get_high_corr = function( substrate, threshold ) {
  sig_list = get_significant_res(result_dir, substrate)
  cor = mapply(return_corr2, as.data.frame(t(COGabun[sig_list,])), substrate)
  high_cor = as.data.frame(t(cor)) %>%
    filter(abs(corr) > threshold) %>%
    arrange(desc(abs(corr)))
  high_cor$COG_function = resolve_COG_function(rownames(high_cor))
  high_cor$substrate = substrate
  print(paste("I have", length(sig_list), "COGs, and", nrow(high_cor), "have correlations greater than", threshold))
  return(high_cor)
}

high_oxygen_cor = get_high_corr("Mean_Oxygen", threshold = 0.6) # 1640 COGs -> 43 high cor
high_phosphate_cor = get_high_corr("log_PO4", threshold = 0.6) # 436 COGs -> 76
high_NO2NO3_cor = get_high_corr("log_NO2NO3", threshold = 0.6) # 549 COGs -> 98

all_cor = rbind(high_oxygen_cor, high_phosphate_cor, high_NO2NO3_cor)
```

# looking at metadata only
```{r}
pairs(~OG.Shannon+log_FC_heterotrophs+log_FC_bacteria+log_NO2NO3+log_PO4+Mean_Oxygen,data=tara_md)

cor.test(tara_md$`FC - bacteria`, tara_md$PO4, method = "spearman")
cor.test(tara_md$`FC - bacteria`, tara_md$NO2NO3, method = "spearman")

cor.test(tara_md$Mean_Oxygen, tara_md$NO2NO3, method = "spearman")
cor.test(tara_md$Mean_Oxygen, tara_md$PO4, method = "spearman")

cor.test(tara_md$OG.Shannon, tara_md$PO4, method = "spearman")
cor.test(tara_md$OG.Shannon, tara_md$Mean_Oxygen, method = "spearman")

plot(NO2NO3~PO4, tara_md)
plot(Mean_Oxygen~PO4, tara_md)
```

# Overlaps between these cOGs
```{r}
all_sig_COGs_df = list(oxygen = get_significant_res(result_dir, "Mean_Oxygen"),
                    NO2NO3 = get_significant_res(result_dir, "log_NO2NO3"),
                    phosphate = get_significant_res(result_dir, "log_PO4"))
ggVennDiagram(all_sig_COGs_df) + 
  scale_color_manual(values = c("gray","gray","gray")) +
  scale_fill_gradient(low = "#F4FAFE", high = "#4981BF")

ggsave("./graphs/VennDiagram_all_COGs.jpg", width = 6, height = 5)

high_cor_COGs_df = list(oxygen = rownames(high_oxygen_cor),
                        NO2NO3 = rownames(high_NO2NO3_cor),
                        phosphate = rownames(high_phosphate_cor))
ggVennDiagram(high_cor_COGs_df) + 
  scale_color_manual(values = c("gray","gray","gray")) +
  scale_fill_gradient(low = "#F4FAFE", high = "#4981BF")

# the two intersects:
# COG2308 3.621703e-28 -0.7811650  Uncharacterized conserved protein
# COG3484 2.346143e-27 -0.7738345 Predicted proteasome-type protease

ggsave("./graphs/VennDiagram_high_cor_COGs.jpg", width = 6, height = 5)
```


# Bray-Curtis PCoA with vegan

PCoA stands for Principal Coordinate Analysis: a sister of Principal
Componenet Analysis. It is better than PCA when there are more features
than samples.

**I promise, the PCA plot tutorial is coming!!!**

Use sqrt instead of log to transform, because PCoA does not take
in negative values in the matrix (computing a negative distance does
not make sense).

```{r}
library(vegan)
all_sig_COGs = unique(rownames(all_cor))
write(all_sig_COGs, file = "high-corr-COGs.txt")
COGabun_sig = COGabun[all_sig_COGs, ]

PCoA_df = t(COGabun_sig)

# https://ordnews.colostate.narkive.com/lMWF502c/1593-log-sqrt-and-other-transformation-with-bray-curtis-dissimilarity
#load dataset
data.bray = vegdist(sqrt(PCoA_df))
#build coordinates
data.b.pcoa = cmdscale(data.bray, k = (nrow(PCoA_df) - 1), eig = TRUE)
#summary
str(data.b.pcoa)
#build df from first two coordinates
pcoa = data.frame(PC1 = data.b.pcoa$points[,1], PC2 = data.b.pcoa$points[,2])

pcoa = merge(pcoa, tara_md, by =0) #work-in metadata

gen_save_PCoA = function(substrate) {
  df = pcoa[c('PC1', 'PC2')]
  df$metadata = as.numeric(pcoa[,substrate])
  ggplot(df, aes(x = PC1, y = PC2, color = metadata)) + 
    geom_point(size=1.8) +
    scale_color_gradient(
      low = "#00B018",
      high = "#CA002A",
      na.value = "grey50",
    ) +
    labs(color=paste(substrate, "(umol/L)")) +
    theme_classic()
  ggsave(paste0("./graphs/PCoA_", substrate, ".png"), 
       plot = last_plot(),
       width = 7,
       height = 6)
}

#viz
gen_save_PCoA("log_NO2NO3")
gen_save_PCoA("log_PO4")
gen_save_PCoA("Mean_Oxygen")
```

# distances between groups of COGs
This is like a dissimilarity distance graph, but with different groups of COGs
```{r}
# COGabun_sig from last chuck

data.bray.2 = vegdist(COGabun_sig)
data.b.pcoa.2 = cmdscale(data.bray.2, k = (nrow(COGabun_sig) - 1), eig = TRUE)
pcoa.2 = data.frame(PC1 = data.b.pcoa.2$points[,1], PC2 = data.b.pcoa.2$points[,2])

# get the respective substrate of each sig COGs
pcoa.2.g = merge(pcoa.2, all_cor, by =0)

ggplot(pcoa.2.g, aes(x = PC1, y = PC2, color = substrate)) + 
    geom_point(aes(size=corr, alpha = 0.5)) +
    theme_classic()
```


# the PCoA doesn't show good results, lemme try PCA
```{r}
COG.pca = prcomp(sqrt(COGabun_sig), scale = T)
out_name = "./graphs/PCA_NO2NO3_PO4_O2.jpg"

# COG.pca = prcomp(sqrt(COGabun[c(rownames(high_NO2NO3_cor), rownames(high_phosphate_cor)), ]))
# out_name = "./graphs/PCA_NO2NO3_PO4.jpg"

COG.pca.coords = as.data.frame(COG.pca$x)[1:2]
COG.pca.coords = merge(COG.pca.coords, all_cor, by =0)
COG.pca.coords$substrate = str_replace(COG.pca.coords$substrate, "log_|Mean_", "")

ggplot(COG.pca.coords, aes(x = PC1, y = PC2, shape = substrate)) + 
    geom_point(aes(color=corr), size = 2) +
    scale_color_gradient(
      low = "#00B018",
      high = "#CA002A",
      na.value = "grey50",
    ) +
    theme_classic()

ggsave(out_name, width = 7, height = 6)
```


# Screen for simple spearman correlation between phosphate/nitrate concentration and gene abundance
```{r}
sig_oxygen = get_significant_res(result_dir, "Mean_Oxygen") # 1395 COGs
sig_phosphate = get_significant_res(result_dir, "log_PO4") # 385 COGs
# sig_nitrates = get_significant_res(result_dir, 'log_Nitrates') # Nitrates = 66
# sig_nitrites = get_significant_res(result_dir, 'log_NO2') # NO2 = 1169
sig_NO2NO3 = get_significant_res(result_dir, 'log_NO2NO3') # NO2NO3 = 205

return_corr2 = function(x, substrate) {
  y = tara_md %>% pull(substrate)
  obj = cor.test(x, y, method = "spearman")
  return(c(p.val = obj$p.value, corr = as.numeric(obj$estimate)))
}

process_corr = function(x, threshold) {
  filter(as.data.frame(t(x)), abs(corr) > threshold) %>% arrange(desc(abs(corr)))
}

oxygen_cor = mapply(return_corr2, as.data.frame(t(COGabun[sig_oxygen,])), "Mean_Oxygen")
high_oxygen_cor = process_corr(oxygen_cor, 0.6) # corr > 0.6: 40 rows

phosphate_cor = mapply(return_corr2, as.data.frame(t(COGabun[sig_phosphate,])), "PO4")
high_phosphate_cor = process_corr(phosphate_cor, 0.6) # corr > 0.6: 66 rows

# nitrate_cor = mapply(return_corr2, as.data.frame(t(COGabun[sig_nitrates,])), "Mean_Nitrates")
# high_nitrate_cor = process_corr(nitrate_cor, 0.6) # corr > 0.6: 13 rows

# nitrite_cor = mapply(return_corr2, as.data.frame(t(COGabun[sig_nitrites,])), "NO2")
# high_nitrite_cor = process_corr(nitrite_cor, 0.6) # corr > 0.6: 3 rows

NO2NO3_cor = mapply(return_corr2, as.data.frame(t(COGabun[sig_NO2NO3,])), "NO2NO3")
high_NO2NO3_cor = process_corr(NO2NO3_cor, 0.6) # corr > 0.6: 12 rows
```


## heatmap!

Focusing on the concentration of phosphorus in the environment.
Try to find out which gene is enriched in phosphorus-poor samples.

I used some functions in *.init_MaAsLin.R* to make this process simpler.

**Warning: phosphate abundance is negatively associated with temperature!!!**

And I don't really know what normalization scheme to make the heat map
more visible (log will still make some gene overly abundant and ruin the graph).
I used a weird 6th square of things.

```{r}
library(pheatmap)

plot(log_phosphate~temperature, BGS_md)
summary(lm(log_phosphate~temperature, BGS_md))

md_to_graph = BGS_md[c("log_phosphate","temperature")]
md_to_graph = md_to_graph[order(md_to_graph$log_phosphate, md_to_graph$temperature),]
nga_g = gen_df_for_graphing(nga, result_dirtory, "log_phosphate")

sign_genes <- get_significant_res(result_dirtory, "log_phosphate", strict = 0.05) 
heatmap_df <- as.data.frame(t(nga[sign_genes]))

g = pheatmap(mat = (heatmap_df[rownames(md_to_graph)])^(1/6),
           annotation_col = md_to_graph,
           cluster_cols = F,
           cluster_rows = F,
           show_rownames = T,
           show_colnames = F,
           scale = "none")

save_pheatmap_pdf(g, "./manual_analysis_graphs/phosphate_sig_genes_heatmap.pdf", height = 5)
```
