---
title: "Tara Ocean metagenome COG and nutrient limitations"
author: "Jimmy Zhong"
date: "4/12/2023"
output: html_document
---

## Jimmy's ibaray set up
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# setting the working directory to the location of this file
setwd(dirname(rstudioapi::getActiveDocumentContext()$path ))
source(".init_tara.R")

tara_md = init_tara_metadata()
COGabun = get_COGabundance() # only the prokaryote fraction

tara_md = tara_md[colnames(COGabun), ]

# result_dir = 'MaAsLin_out/logPO4_Salinity_Oxygen_Nitrogen_temp_logDepth_size'
# param = c('size_fraction','Mean_Temperature', "log_depth", 'Mean_Salinity',
#          'Mean_Oxygen', 'log_PO4', 'log_Nitrates', 'log_NO2', 'log_NO2NO3')

# result_dir = 'MaAsLin_out/logPO4_logNO2NO3_Salinity_Oxygen_temp_logDepth_size'
# param = c('size_fraction','Mean_Temperature', "log_depth", 'Mean_Salinity',
#           'Mean_Oxygen', 'log_PO4','log_NO2NO3')

result_dir = 'MaAsLin_out/logPO4_sqrtIron_logNO2NO3_Salinity_Oxygen_temp_logDepth_size'
param = c('size_fraction','Mean_Temperature', "log_depth", 'Mean_Salinity',
           'Mean_Oxygen', 'log_PO4','log_NO2NO3', "sqrt_iron")
```

# run MaAsLin2
```{r}
fit_data <- Maaslin2(
    COGabun, tara_md, result_dir, transform = "LOG",
    fixed_effects = param,
    random_effects = c("Mean_Lat", "Mean_Long"),
    reference = 'size_fraction,0.22-1.6',
    max_significance = 0.05,
    normalization = 'NONE',
    min_prevalence = 0.33, # at least 1/3 of the sample has it
    standardize = FALSE)
```

# Screen for simple spearman correlation between phosphate/nitrate concentration and gene abundance
```{r}
return_corr2 = function(x, substrate, mthd = "spearman") {
  y = tara_md %>% pull(substrate)
  obj = cor.test(x, y, method = mthd)
  return(c(p.val = obj$p.value, corr = as.numeric(obj$estimate)))
}

get_high_corr = function( substrate, threshold ) {
  sig_list = get_significant_res(result_dir, substrate)
  cor = mapply(return_corr2, as.data.frame(t(COGabun[sig_list,])), substrate)
  high_cor = as.data.frame(t(cor)) %>%
    filter(abs(corr) > threshold) %>%
    arrange(desc(abs(corr))) %>%
    tibble::rownames_to_column("COG")
  high_cor$COG_function = resolve_COG_function(high_cor$COG)
  high_cor$substrate = substrate
  print(paste("I have", length(sig_list), "COGs, and", nrow(high_cor), "have correlations greater than", threshold))
  return(high_cor)
}

high_oxygen_cor = get_high_corr("Mean_Oxygen", threshold = 0.6) # with iron: 522 -> 21; no iron: 1640 COGs -> 43 high cor
high_phosphate_cor = get_high_corr("log_PO4", threshold = 0.6) # with iron: 282 -> 73; no iron: 436 COGs -> 76
high_NO2NO3_cor = get_high_corr("log_NO2NO3", threshold = 0.6) #with iron: 273 -> 48; no iron: 549 COGs -> 98
high_iron_cor = get_high_corr("sqrt_iron", threshold = 0.6) # 501 COGs -> 12

iron_text_COGs=c("COG3470","COG0370","COG0672","COG0735","COG1814","COG1914","COG1918","COG4592","COG1120","COG3842","COG0609","COG0614","COG1178","COG1840","COG4594","COG4772")

iron_text_COGs_corr = high_iron_cor[high_iron_cor$COG %in% iron_text_COGs, ]
all_cor = rbind(high_NO2NO3_cor, high_oxygen_cor, high_phosphate_cor, high_iron_cor)
write.csv(all_cor, "high-corr-COGs.csv", row.names=FALSE)

split_dir = function(df) { return(list(df[df$corr >0, ], df[df$corr <0, ])) }
g(pos_oxygen, neg_oxygen) %=% split_dir(high_oxygen_cor)
g(pos_phosph, neg_phosph) %=% split_dir(high_phosphate_cor)
g(pos_iron, neg_iron) %=% split_dir(high_iron_cor)
g(pos_NO2NO3, neg_NO2NO3) %=% split_dir(high_NO2NO3_cor)

all_dir_sum = t(data.frame(
  pos_oxygen = paste(pos_oxygen$COG, collapse = " "),
  neg_oxygen = paste(neg_oxygen$COG, collapse = " "),
  pos_phosphate = paste(pos_phosph$COG, collapse = " "),
  neg_phosphate = paste(neg_phosph$COG, collapse = " "),
  pos_iron = paste(pos_iron$COG, collapse = " "),
  neg_iron = paste(neg_iron$COG, collapse = " "),
  pos_NO2NO3 = paste(pos_NO2NO3$COG, collapse = " "),
  neg_NO2NO3 = paste(neg_NO2NO3$COG, collapse = " ")
))
write.csv(all_dir_sum, "pos_neg_separated_COGs.csv", row.names=TRUE)
```

# looking at metadata only
```{r}
pairs(~OG.Shannon+log_FC_heterotrophs+log_FC_bacteria+log_NO2NO3+log_PO4+Mean_Oxygen,data=tara_md)

cor.test(tara_md$`FC - bacteria`, tara_md$PO4, method = "spearman")
cor.test(tara_md$`FC - bacteria`, tara_md$NO2NO3, method = "spearman")

cor.test(tara_md$Mean_Oxygen, tara_md$NO2NO3, method = "spearman")
cor.test(tara_md$Mean_Oxygen, tara_md$PO4, method = "spearman")

cor.test(tara_md$OG.Shannon, tara_md$PO4, method = "spearman")
cor.test(tara_md$OG.Shannon, tara_md$Mean_Oxygen, method = "spearman")

tmp = tara_md[!is.na(tara_md$NO2),]
tmp = tmp[!is.na(tmp$NO2NO3), ]
tmp$t = tmp$NO2NO3 - tmp$NO2
tmp$`NO2NO3-NO2` = ifelse(tmp$t < 0, 0, tmp$t)

plot(log(NO2 + 0.001)~log(`NO2NO3-NO2` + 0.001), tmp)


plot(NO2NO3~PO4, tara_md)
plot(Mean_Oxygen~PO4, tara_md)
plot(log_PO4~log_depth, tara_md)
boxplot(log_PO4~layer, tara_md)
```

# Overlaps between these cOGs
```{r}
four_gray = rep("gray", 4)

all_sig_COGs_df = list(oxygen = get_significant_res(result_dir, "Mean_Oxygen"),
                    NO2NO3 = get_significant_res(result_dir, "log_NO2NO3"),
                    phosphate = get_significant_res(result_dir, "log_PO4"),
                    iron = get_significant_res(result_dir, "sqrt_iron"))
ggVennDiagram(all_sig_COGs_df) + 
  scale_color_manual(values = four_gray) +
  scale_fill_gradient(low = "#F4FAFE", high = "#4981BF")

ggsave("./graphs/VennDiagram_all_COGs_iron.jpg", width = 6, height = 5)

high_cor_COGs_df = list(oxygen = rownames(high_oxygen_cor),
                        NO2NO3 = rownames(high_NO2NO3_cor),
                        phosphate = rownames(high_phosphate_cor),
                        iron = rownames(high_iron_cor))
ggVennDiagram(high_cor_COGs_df) + 
  scale_color_manual(values = four_gray) +
  scale_fill_gradient(low = "#F4FAFE", high = "#4981BF")

# the two intersects:
# COG2308 3.621703e-28 -0.7811650  Uncharacterized conserved protein
# COG3484 2.346143e-27 -0.7738345 Predicted proteasome-type protease

ggsave("./graphs/VennDiagram_high_cor_COGs.jpg", width = 6, height = 5)
```


# Bray-Curtis PCoA with vegan

PCoA stands for Principal Coordinate Analysis: a sister of Principal
Componenet Analysis. It is better than PCA when there are more features
than samples.

**I promise, the PCA plot tutorial is coming!!!**

Use sqrt instead of log to transform, because PCoA does not take
in negative values in the matrix (computing a negative distance does
not make sense).

```{r}
library(vegan)
all_sig_COGs = unique(all_cor$COG)
write(all_sig_COGs, file = "high-corr-COGs.txt")
COGabun_sig = COGabun[all_sig_COGs, ]

PCoA_df = t(COGabun_sig)

# https://ordnews.colostate.narkive.com/lMWF502c/1593-log-sqrt-and-other-transformation-with-bray-curtis-dissimilarity
#load dataset
data.bray = vegdist(sqrt(PCoA_df))
#build coordinates
data.b.pcoa = cmdscale(data.bray, k = (nrow(PCoA_df) - 1), eig = TRUE)
#summary
str(data.b.pcoa)
#build df from first two coordinates
pcoa = data.frame(PC1 = data.b.pcoa$points[,1], PC2 = data.b.pcoa$points[,2])

pcoa = merge(pcoa, tara_md, by =0) #work-in metadata

gen_save_PCoA = function(substrate) {
  df = pcoa[c('PC1', 'PC2')]
  df$metadata = as.numeric(pcoa[,substrate])
  ggplot(df, aes(x = PC1, y = PC2, color = metadata)) + 
    geom_point(size=1.8) +
    scale_color_gradient(
      low = "#00B018",
      high = "#CA002A",
      na.value = "grey50",
    ) +
    labs(color=paste(substrate, "(umol/L)")) +
    theme_classic()
  ggsave(paste0("./graphs/PCoA_", substrate, ".png"), 
       plot = last_plot(),
       width = 7,
       height = 6)
}

#viz
gen_save_PCoA("log_NO2NO3")
gen_save_PCoA("log_PO4")
gen_save_PCoA("Mean_Oxygen")
gen_save_PCoA("sqrt_iron")
```

# distances between groups of COGs
This is like a dissimilarity distance graph, but with different groups of COGs
```{r}
# COGabun_sig from last chuck

data.bray.2 = vegdist(COGabun_sig)
data.b.pcoa.2 = cmdscale(data.bray.2, k = (nrow(COGabun_sig) - 1), eig = TRUE)
pcoa.2 = data.frame(PC1 = data.b.pcoa.2$points[,1], PC2 = data.b.pcoa.2$points[,2])

# get the respective substrate of each sig COGs
pcoa.2.g = merge(pcoa.2, all_cor, by =0)

ggplot(pcoa.2.g, aes(x = PC1, y = PC2, color = substrate)) + 
    geom_point(aes(size=corr, alpha = 0.5)) +
    theme_classic()
```


# the PCoA doesn't show good results, lemme try PCA
```{r}
COG.pca = prcomp(sqrt(COGabun_sig), scale = T)
out_name = "./graphs/PCA_NO2NO3_PO4_O2.jpg"

# COG.pca = prcomp(sqrt(COGabun[c(rownames(high_NO2NO3_cor), rownames(high_phosphate_cor)), ]))
# out_name = "./graphs/PCA_NO2NO3_PO4.jpg"

COG.pca.coords = as.data.frame(COG.pca$x)[1:2]
COG.pca.coords = merge(COG.pca.coords, all_cor, by =0)
COG.pca.coords$substrate = str_replace(COG.pca.coords$substrate, "log_|Mean_", "")

ggplot(COG.pca.coords, aes(x = PC1, y = PC2, shape = substrate)) + 
    geom_point(aes(color=corr), size = 2) +
    scale_color_gradient(
      low = "#00B018",
      high = "#CA002A",
      na.value = "grey50",
    ) +
    theme_classic()

ggsave(out_name, width = 7, height = 6)
```


# Screen for simple spearman correlation between phosphate/nitrate concentration and gene abundance
```{r}
sig_oxygen = get_significant_res(result_dir, "Mean_Oxygen") # 1395 COGs
sig_phosphate = get_significant_res(result_dir, "log_PO4") # 385 COGs
# sig_nitrates = get_significant_res(result_dir, 'log_Nitrates') # Nitrates = 66
# sig_nitrites = get_significant_res(result_dir, 'log_NO2') # NO2 = 1169
sig_NO2NO3 = get_significant_res(result_dir, 'log_NO2NO3') # NO2NO3 = 205

return_corr2 = function(x, substrate) {
  y = tara_md %>% pull(substrate)
  obj = cor.test(x, y, method = "spearman")
  return(c(p.val = obj$p.value, corr = as.numeric(obj$estimate)))
}

process_corr = function(x, threshold) {
  filter(as.data.frame(t(x)), abs(corr) > threshold) %>% arrange(desc(abs(corr)))
}

oxygen_cor = mapply(return_corr2, as.data.frame(t(COGabun[sig_oxygen,])), "Mean_Oxygen")
high_oxygen_cor = process_corr(oxygen_cor, 0.6) # corr > 0.6: 40 rows

phosphate_cor = mapply(return_corr2, as.data.frame(t(COGabun[sig_phosphate,])), "PO4")
high_phosphate_cor = process_corr(phosphate_cor, 0.6) # corr > 0.6: 66 rows

# nitrate_cor = mapply(return_corr2, as.data.frame(t(COGabun[sig_nitrates,])), "Mean_Nitrates")
# high_nitrate_cor = process_corr(nitrate_cor, 0.6) # corr > 0.6: 13 rows

# nitrite_cor = mapply(return_corr2, as.data.frame(t(COGabun[sig_nitrites,])), "NO2")
# high_nitrite_cor = process_corr(nitrite_cor, 0.6) # corr > 0.6: 3 rows

NO2NO3_cor = mapply(return_corr2, as.data.frame(t(COGabun[sig_NO2NO3,])), "NO2NO3")
high_NO2NO3_cor = process_corr(NO2NO3_cor, 0.6) # corr > 0.6: 12 rows
```
