---
title: "Tara Ocean metagenome COG and nutrient limitations"
author: "Jimmy Zhong"
date: "4/12/2023"
output: html_document
---

## Jimmy's ibaray set up
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# setting the working directory to the location of this file
setwd(dirname(rstudioapi::getActiveDocumentContext()$path ))
source(".init_tara.R")

tara_md = init_tara_metadata()
COGabun = get_COGabundance() # only the prokaryote fraction

tara_md = tara_md[colnames(COGabun), ]
tara_md$size_fraction = as.factor(tara_md$size_fraction)
```

# run MaAsLin2
```{r}
result_dirtory = 'MaAsLin_out/logPO4_Salinity_Oxygen_Nitrogen_temp_logDepth_size'

fit_data <- Maaslin2(
    COGabun, tara_md, result_dirtory, transform = "LOG",
    fixed_effects = c('size_fraction','Mean_Temperature', "log_depth",
                      'Mean_Salinity', 'Mean_Oxygen', 'log_PO4',
                      'log_Nitrates', 'log_NO2', 'log_NO2NO3'),
    random_effects = c("Mean_Lat", "Mean_Long"),
    reference = 'size_fraction,0.22-1.6',
    max_significance = 0.05,
    normalization = 'NONE',
    min_prevalence = 0.33, # at least 1/3 of the sample has it
    standardize = FALSE)

tmp = COGabun[c("COG0226", "COG0573", "COG0581", "COG4985", "COG1117", "COG3638", "COG3221", "COG3639", "COG3624", "COG3625", "COG3626", "COG3627", "COG4107", "COG4778", "COG3454", "COG3709", "COG2764", "COG0395", "COG1653", "COG1134", "COG1392", "COG1840", "COG0704"), ]
tmp$COG_function = c("PstS", "PstC", "PstA", "PstA_auxiliary", "PstB", "PhnC", "PhnD", "PhnE", "PhnG", "PhnH", "PhnI", "PhnJ", "PhnK", "PhnL", "PhnM", "PhnN", "PhnB", "UgpE", "UgpB", "TagH", "YkaA", "AfuA", "PhoU")
tmp = tmp[c(ncol(tmp), 2:ncol(tmp)-1)]

write.csv(tmp, "phosphorous_COGs_abundance.csv", row.names=TRUE)
```

# Bray-Curtis PCoA with vegan

PCoA stands for Principal Coordinate Analysis: a sister of Principal
Componenet Analysis. It is better than PCA when there are more features
than samples.

**I promise, the PCA plot tutorial is coming!!!**

Use sqrt instead of log to transform, because PCoA does not take
in negative values in the matrix (computing a negative distance does
not make sense).

```{r}
library(vegan)
# get_significant_res(): Jimmy's function in .init_MaAsLin.R.
# strict: set significant level
all_sign_genes <- get_significant_res(result_dirtory, "log_phosphate", strict = 0.05) 
PCoA_df <- nga[all_sign_genes]

# https://ordnews.colostate.narkive.com/lMWF502c/1593-log-sqrt-and-other-transformation-with-bray-curtis-dissimilarity
#load dataset
data.bray = vegdist(sqrt(PCoA_df))
#build coordinates
data.b.pcoa = cmdscale(data.bray, k = (nrow(PCoA_df) - 1), eig = TRUE)
#summary
str(data.b.pcoa)
#build df from first two coordinates
pcoa = data.frame(PC1 = data.b.pcoa$points[,1], PC2 = data.b.pcoa$points[,2])

pcoa = merge(pcoa, BGS_md, by =0) #work-in metadata

#viz
ggplot(pcoa, aes(x = PC1, y = PC2, color = log_phosphate)) + 
  geom_point(size=1.8)
ggsave("./manual_analysis_graphs/PCoA_log_phosphate.png", 
       plot = last_plot(),
       width = 9,
       height = 6)
```


## heatmap!

Focusing on the concentration of phosphorus in the environment.
Try to find out which gene is enriched in phosphorus-poor samples.

I used some functions in *.init_MaAsLin.R* to make this process simpler.

**Warning: phosphate abundance is negatively associated with temperature!!!**

And I don't really know what normalization scheme to make the heat map
more visible (log will still make some gene overly abundant and ruin the graph).
I used a weird 6th square of things.

```{r}
library(pheatmap)

plot(log_phosphate~temperature, BGS_md)
summary(lm(log_phosphate~temperature, BGS_md))

md_to_graph = BGS_md[c("log_phosphate","temperature")]
md_to_graph = md_to_graph[order(md_to_graph$log_phosphate, md_to_graph$temperature),]
nga_g = gen_df_for_graphing(nga, result_dirtory, "log_phosphate")

sign_genes <- get_significant_res(result_dirtory, "log_phosphate", strict = 0.05) 
heatmap_df <- as.data.frame(t(nga[sign_genes]))

g = pheatmap(mat = (heatmap_df[rownames(md_to_graph)])^(1/6),
           annotation_col = md_to_graph,
           cluster_cols = F,
           cluster_rows = F,
           show_rownames = T,
           show_colnames = F,
           scale = "none")

save_pheatmap_pdf(g, "./manual_analysis_graphs/phosphate_sig_genes_heatmap.pdf", height = 5)
```
